DOCSP-30548: monitoring (command, connection pool, cluster topology) (#69)
mongodb/docs-rust: source/fundamentals/monitoring/connection-monitoring.txt

.. _rust-connection-monitoring:

==========================
Connection Pool Monitoring
==========================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecols

Overview
--------

This guide shows you how to use the {+driver-short+} to monitor the

> driver's **connection pool**. A connection pool is a set of open TCP

Reviewer: S: not necessary, but helpful to spell out the full acronym first
```suggestion
driver's **connection pool**. A connection pool is a set of open Transmission Control Protocol (TCP)
```

> connections your driver maintains with a MongoDB instance. Connection

Reviewer: S: 
```suggestion
connections that your driver maintains with a MongoDB instance. Connection
```

> pools help reduce the number of network handshakes your application

Reviewer: S: "network handshakes" might qualify as [jargon to avoid](https://www.mongodb.com/docs/meta/style-guide/terminology/terms-for-global-audience/#use-jargon-with-care); maybe say something like "number of times new connections are created" instead?

needs to perform, which might make your application run faster.

You can use information about changes to your connection pool in your
application, or you can monitor the connection pool to learn more about
how the driver uses connections.

Event Subscription Example
--------------------------

You can access one or more connection pool events by
subscribing to them in your application. The following example connects
to a MongoDB deployment and subscribes to the ``ConnectionCreatedEvent`` event type:

.. literalinclude:: /includes/fundamentals/code-snippets/monitoring/cpm-subscribe.rs
   :language: rust
   :dedent:
   :start-after: begin-cmap
   :end-before: end-cmap

> Event Descriptions

Reviewer: S: same suggestion about reordering sections as above

------------------

You can subscribe to one or more of the following connection pool
monitoring events:

.. list-table::
   :widths: 33 67
   :header-rows: 1

   * - Event Name
     - Description

   * - ``PoolCreatedEvent``
     - Created when a connection pool is created.

   * - ``PoolReadyEvent``
     - Created when a connection pool is ready.

   * - ``PoolClearedEvent``
     - Created when a connection pool is cleared.

   * - ``PoolClosedEvent``
     - Created when a connection pool is closed, before the destruction of
       the server instance.

   * - ``ConnectionCreatedEvent``
     - Created when a connection is created, but not necessarily
       when it is used for an operation.

   * - ``ConnectionReadyEvent``
     - Created after a connection successfully completes a
       handshake and is ready to be used for operations.

   * - ``ConnectionClosedEvent``
     - Created when a connection is closed.

   * - ``ConnectionCheckoutStartedEvent``
     - Created when an operation attempts to acquire a connection for
       execution.

   * - ``ConnectionCheckoutFailedEvent``

>      - Created when an operation fails to acquire a connection for

Reviewer: S: to avoid "fails"

```suggestion
     - Created when an operation cannot acquire a connection for
```

.. _rust-connection-monitoring:

==========================
Connection Pool Monitoring
==========================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecols

Overview
--------

This guide shows you how to use the {+driver-short+} to monitor the

> driver's **connection pool**. A connection pool is a set of open

Reviewer: S: nit, but since "cluster" and "command" aren't bold in the other two monitoring pages, I feel like "connection pool" shouldn't be. I'm actually not sure what the general rule is for bolding terms, so ignore if you want to keep it bold

     address: ...,
     reason: ...,
     /* private fields */
   }

ConnectionCheckedOutEvent
~~~~~~~~~~~~~~~~~~~~~~~~~

.. code-block:: none
   :copyable: false

   ConnectionCheckedOutEvent {
     address: ...,
     connection_id: 1
   }

ConnectionCheckedInEvent
~~~~~~~~~~~~~~~~~~~~~~~~

.. code-block:: none
   :copyable: false

   ConnectionCheckedInEvent {
     address: ...,
     connection_id: 1
   }

.. _rust-cpm-addtl-info:

Additional Information
----------------------

To learn more about monitoring a MongoDB deployment, see the :website:`How
to Monitor MongoDB
</basics/how-to-monitor-mongodb-and-what-metrics-to-monitor>` article.

To learn more about connecting to MongoDB, see the
:ref:`rust-connect-to-mongodb`.

To learn more about improving the performance of your application, see
the guide on :ref:`rust-performance`.

API Documentation
~~~~~~~~~~~~~~~~~

To learn more about the methods and types mentioned in this
guide, see the following API documentation:

- `CmapEventHandler <{+api+}/event/cmap/trait.CmapEventHandler.html>`__
- `handle_connection_created_event() <{+api+}/event/cmap/trait.CmapEventHandler.html#method.handle_connection_created_event>`__

> - `ConnectionCreatedEvent <{+api+}/event/cmap/struct.ConnectionCreatedEvent.html>`__

Reviewer: S: same suggestion as my comment on cluster monitoring (line 238)
